### ISSUE TITLE
Move PDF creation + emailing into background job

### ISSUE THREAD

**Issue by daguar** [codeforamerica/clean#349]
Currently it takes 7-9 seconds when the user clicks "submit" for them to get a response that the application has been sent in. That's because writing to a PDF is a pretty time-intensive operation.

Rather than trying to do incremental optimizations on that ~7 second number or further UX hacks to make the user happier waiting (reasonable at prototype stage), the better long-term solution given our focus on Clean is to invest in the work to move this to a background job.

Because we're now on Rails 4.2, we also have a nice infrastructure for this — we can use the nifty new ActiveJob ( http://edgeguides.rubyonrails.org/active_job_basics.html ) queue library.

It should in fact be able to execute immediately in the background in parallel with responding to the web request immediately (the default behavior in ActiveJob, I believe.)

**Comment by daguar** [comment]
Note: we'll have to figure out how best to deal with failure cases. This should be reasonably doable except for edge cases, but just merits a bit of thoughtfulness, since we would no longer know immediately if writing to PDF failed. The basic solution would likely just be validation of user input to ensure it's in a state we trust _before_ we send it off to a PDF writing job.

**Comment by daguar** [comment]
This is basically implemented, but needs a rebase.

**Comment by alanjosephwilliams** [comment]
from our IRL conversation, it sounds like the impact on the user's
experience will be narrow, but significant. I've observed (and gotten
questions, in the moment) anxiety from assisters—most memorably,
Francesca—when submitting currently.

While we don't have an open for *submission*, we have documented that we
are not giving strong enough visual cues when uploading documents.
Submission is a similar experience—our button copy and state changes, but
the progress of the submission is shown only in the default browser
progress bar.

If I understand correctly, users will now experience their application
being submitted instantaneously, on par with the experience of loading a
new page. This is a big win.

On Tue, Feb 24, 2015 at 7:22 AM, Dave Guarino <notifications@github.com>
wrote:

> This is basically implemented, but needs a rebase.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/codeforamerica/clean/issues/349#issuecomment-75776177>
> .
>



-- 
t:  @alanjosephwilli
p: 817 713 6264

**Comment by daguar** [comment]
> If I understand correctly, users will now experience their application being submitted instantaneously, on par with the experience of loading a new page.

Correct. Let's dooooo it! (I've moved this to current sprint, but let's discuss in sprint planning.)
